### PR TITLE
[1/2] Feature: Integrate Engram to SGLang

### PR DESCRIPTION
**Pull request by magicYang1573**
## Motivation

[Engram](https://arxiv.org/pdf/2601.07372) is a conditional memory mechanism for LLMs that augments Transformer inference with n-gram hash-based embedding lookups. Its massive embedding tables exhibit sparse access patterns and are well-suited for offloading to lower-tier memory.

This PR integrates the Engram module into SGLang, with a clean separation between the model-computation side and the storage side. It also introduces prototype support for Qwen2 and Qwen3 models with Engram enabled (currently implements the architectural compute flow rather than a full model for performance simulation and infrastructure validation). It provides pluggable storage backends to store Engram embeddings in local DRAM.

Relates to #19907

This PR provides the initial infrastructure for Engram in SGLang. The current implementation performs the computation part of Engram primarily on the GPU, while storing the massive embedding tables in DRAM.

### New: `python/sglang/srt/mem_cache/engram/` — storage side

A new package that manages Engram embedding tables independently of the KV cache. Engram stores are **not** KV cache backends; they run alongside whatever HiCache backend is configured.

- **`engram_store.py`** — `EngramStore` abstract base class (`put_sharded`, `get_one`, `get_many`, `close`) and `EngramStoreConfig`.
- **`local_engram_store.py`** — Default backend: CPU DRAM tensors, no external dependencies.
- **`engram_store_manager.py`** — `EngramStoreManager`: a per-process registry and lazy factory for per-layer `EngramStore` instances, with global singleton accessors (`get/set/close_global_engram_store_manager`).

### New: `python/sglang/srt/models/engram/` — computation side

- **`engram.py`** — Core Engram module: `NgramHashMapping` (prime-modulo n-gram hashing, NumPy + Torch paths), `MultiHeadEmbedding` (multi-head embedding lookup via `EngramStore`), `Engram` (gating, value projection, short depthwise convolution, async CUDA-stream prefetch), and `ShortConv`. Also defines `EngramConfig` and `BackBoneConfig` global config objects.
- **`triton_ops/engram_triton.py`** — Optional fused Triton kernels for gate-value computation, grouped RMSNorm, and short-conv preprocessing (activated via `ENGRAM_TRITON=1`).

### Modified: `python/sglang/srt/models/qwen2.py`

- Added `Qwen2MoelEngram` — a `Qwen2Model` subclass that injects `Engram` modules at configured layer IDs, initialises the global `EngramStoreManager`, and runs async embedding prefetch before each forward pass.

### Modified: `python/sglang/srt/models/qwen3.py`

- Added `Qwen3ModelEngram` (inherits `Qwen2MoelEngram`) and `Qwen3ForCausalLM` opt-in via `ENABLE_ENGRAM=1|true`.

### New: `test/srt/engram/`

Three integration tests runnable via `python test/srt/engram/run_all_tests.py`:
1. **`test_import_paths.py`** — verifies all module import paths are resolvable.
2. **`test_local_store_manager.py`** — tests `LocalEngramStore` lifecycle through `EngramStoreManager` (put/get correctness, idempotency, global accessor).
3. **`test_engram_e2e.py`** — end-to-end forward pass with `EngramStoreManager`, backward compatibility (no manager), and multi-layer scenarios.

### New: `python/sglang/srt/mem_cache/engram/README.md`

Documents architecture, the storage backend with environment variable references and launch commands, quick start, configuration table, and a TODO list.

## Accuracy Tests

This PR introduces a new module that runs in parallel with the existing forward pass. The Engram output is currently added with a zero-coefficient (`engram_output.sum() * 0`) so it does not affect model logits. Existing model accuracy is therefore unchanged.

Integration correctness is verified by `test/srt/engram/run_all_tests.py`, which covers store put/get round-trips and end-to-end `Engram` forward pass shape correctness.

## Benchmarking and Profiling
**Experimental Setup:** Single L20 GPU, Batch Size = 256, Sequence Length = 512 tokens.

| Model | Baseline (tokens/s) | +Engram (DRAM) (tokens/s) | Performance Delta |
| :--- | :--- | :--- | :--- |
| **Qwen3-4B** | 6183.9 | 5683.7 | -8.09% |
| **Qwen3-8B** | 4185.6 | 3909.7 | -6.59% |

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests). (`test/srt/engram/`)
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations). (`mem_cache/engram/README.md`)
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

### CODE REVIEW

File: python/sglang/srt/models/engram/engram.py [hunk 1]
```diff
@@ -0,0 +1,699 @@
...
+        return embeddings
+
+    def parallel_rms_norm(self, x, weight):
+        # x shape: (batch, seq, hc_mult, hidden_size)
+        # weight shape: (hc_mult, hidden_size)
+        variance = x.pow(2).mean(-1, keepdim=True)
+        x = x * torch.rsqrt(variance + self.eps)
+        return x * weight
+
+    def forward(self, hidden_states, input_ids):
```
Comment (CaoE) [inline python/sglang/srt/models/engram/engram.py]: Can we further subdivide the engram to adapt to different device backends? For example, could we abstract the hash and embedding parts into one method; abstract the pre-computed key/value projection and norm parts that don't depend on the query into another method; and abstract the convolution and residual parts into another method, as shown in https://github.com/sgl-project/sglang/issues/19907. This would allow different device backends to provide their own optimization methods and make the scheduling strategy for hybrid execution more flexible.

Comment (magicYang1573) [reply python/sglang/srt/models/engram/engram.py]: We have further abstracted the Engram forward process into three stages
- compute_embeddings: The computation of hash and embedding lookups.
- compute_kv_projections: The key/value projections and normalization.
- compute_mixing: The convolution and gating components.
